tests/unit/helpers.py

```diff
@@ -0,0 +1,30 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""helpers for the unit test."""
+
+import contextlib
+import typing
+import unittest.mock
+
+import ops
+
+
+@contextlib.contextmanager
+def patch_exec(fail: bool = False) -> typing.Generator[unittest.mock.Mock, None, None]:
```

Comment:
Isn't there a new way of doing this now where we don't have to mock ourselves?

Reply:
Check with @weiiwang01 who helped develop it

Reply:
I think that you're talking about [handle_exec](https://ops.readthedocs.io/en/latest/#ops.testing.Harness.handle_exec). But `handle_exec` gives us a way to intercept a call to do something else instead, not test that the call was made.  

I tried to use `handle_exec` with a handler that would raise an Exception, like that:
```
        def test_handler(args):
            raise Exception

        harness.handle_exec(
            "discourse", 
            [],
            handler=test_handler
        )
```

But the test didn't fail. So I'm not sure that I can use `handle_exec` here.

Reply:
Perhaps discuss with @weiiwang01 to see if he has an idea of how to do this?

Reply:
Yes, we have talked, the reason might be that the current mocks intercepted the `ops.testing` calls.

Reply:
After discussing with Weii, we saw that I forgot to remove the patch to `Container.exec` during my test of `handle_exec`. This caused it to catch the exec call instead of `handle_exec`.  
I rewrote the tests using `handle_exec`.